domain/secret/state/state.go

```diff
@@ -0,0 +1,448 @@
+	}
+
+	secretRevisions, err := dbSecretRevisions.toSecretRevisions()
+	if err != nil {
+		return nil, errors.Annotatef(err, "composing secret revision %d for secret %q from database", revision, uri)
+	}
+	return secretRevisions[0], nil
+}
+
+// GetSecretValue returns the contents - either data or value reference - of a given secret revision,
+// returning an error satisfying [secreterrors.SecretRevisionNotFound] if the secret revision does not exist.
+func (st State) GetSecretValue(ctx context.Context, uri *coresecrets.URI, revision int) (coresecrets.SecretData, *coresecrets.ValueRef, error) {
+	db, err := st.DB()
+	if err != nil {
+		return nil, nil, errors.Trace(err)
+	}
+
+	// We look for either content or a value reference, which ever is present.
+	contentQuery := `
+SELECT (*) AS (&secretContent.*)
+FROM   secret_content sc
+JOIN   secret_revision rev ON sc.revision_uuid = rev.uuid  
+WHERE  rev.secret_id = $secretRevision.secret_id AND rev.revision = $secretRevision.revision
+`
+
+	contentQueryStmt, err := st.Prepare(contentQuery, secretContent{}, secretRevision{})
+	if err != nil {
+		return nil, nil, errors.Trace(err)
+	}
+
+	valueRefQuery := `
+SELECT (*) AS (&secretValueRef.*)
+FROM   secret_value_ref val
+JOIN   secret_revision rev ON val.revision_uuid = rev.uuid  
+WHERE  rev.secret_id = $secretRevision.secret_id AND rev.revision = $secretRevision.revision
+`
+
+	valueRefQueryStmt, err := st.Prepare(valueRefQuery, secretValueRef{}, secretRevision{})
+	if err != nil {
+		return nil, nil, errors.Trace(err)
+	}
+
+	want := secretRevision{SecretID: uri.ID, Revision: revision}
+
+	var (
+		dbSecretValues    secretValues
+		dbSecretValueRefs []secretValueRef
+	)
+	if err := db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := tx.Query(ctx, contentQueryStmt, want).GetAll(&dbSecretValues)
+		if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+			return errors.Annotatef(err, "retrieving secret value for %q revision %d", uri, revision)
+		}
+		// Do we have content from the db?
+		if len(dbSecretValues) > 0 {
+			return nil
+		}
+
+		// No content, try a value reference.
+		err = tx.Query(ctx, valueRefQueryStmt, want).GetAll(&dbSecretValueRefs)
+		if errors.Is(err, sqlair.ErrNoRows) {
+			return nil
+		}
+		return errors.Annotatef(err, "rrrrretrieving secret value ref for %q revision %d", uri, revision)
+	}); err != nil {
+		return nil, nil, errors.Annotate(err, "querying secret value")
+	}
+
+	// Compose and return any secret content from the db.
+	if len(dbSecretValues) > 0 {
+		content, err := dbSecretValues.toSecretData()
+		if err != nil {
+			return nil, nil, errors.Annotatef(err, "composing secret contentfor secret %q revision %d from database", uri, revision)
+		}
+		return content, nil, nil
+	}
+
+	// Process any value reference.
+	if len(dbSecretValueRefs) == 0 {
+		return nil, nil, fmt.Errorf("secret value for %q revision %d not found%w", uri, revision, errors.Hide(secreterrors.SecretRevisionNotFound))
```

Comment:
"secret value ref for ..."
And better do "%w: ..."?

Reply:
We've been doing "blah not found" since this is a proper english sentence. Previously we could just append the NotFound error string at the end, but now the error string doesn't create a proper english sentence. So we have to do the %w and Hide() approach.

Reply:
Still have to fix the "secret value `ref` for" tho.

Reply:
Ah doh. Yes